### PR TITLE
Add CRC32 checkSum control

### PR DESCRIPTION
**Pull request by Gsantomaggio**
- Small fix application properties in case of byte value on reading.
- Closes https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/issues/19
- possibility to enable/disable the CRC32 control with the `Crc32` interface 
- The user is free to use the Crc implementation

RawConsumer: 
  ```csharp
          
        private readonly ICrc32 _crc32 = new MyCrcImplentation()
          new RawConsumerConfig(stream)
                {
                    Crc32 = _crc32,
                
   ```

By default, the CRC control is **disabled** to be compatible with the correct version. 
Enable the CRC control could reduce the performances in some cases since it has to calculate the CRC for each chunk.

It is recommended to enable it. 

The documentation needs to be included; once this is approved, we will update the documentation.

### CODE REVIEW

**Comment by jonnepmyra** [comment]
Currently evaluating this PR in our dev environments, looks good so far, no unexpected behaviours found, will continue to monitor for a couple of days

**Comment by michaelklishin** [comment]
@jonnepmyra thank you for following up, we really appreciate the feedback!

File: RabbitMQ.Stream.Client/Deliver.cs [hunk 1]
```diff
@@ -93,8 +93,13 @@ internal static int Read(ref SequenceReader<byte> reader, byte entryType, out Su
             // We need to pass it to the subEntryChunk that will decode the information
             var memory =
                 ArrayPool<byte>.Shared.Rent((int)dataLen).AsMemory(0, (int)dataLen);
-            var data = reader.Sequence.Slice(reader.Consumed, dataLen);
-            data.CopyTo(memory.Span);
+
+            //see the Chunk.Read for more details about this check
+            if (!reader.TryCopyTo(memory.Span))
```
Comment (Gsantomaggio) [inline RabbitMQ.Stream.Client/Deliver.cs:98]:
```
   var data = reader.Sequence.Slice(reader.Consumed, dataLen);
   data.CopyTo(memory.Span);
```
was removed in favour of:
```
   if (!reader.TryCopyTo(memory.Span))
```

it is faster since avoids a  `eader.Sequence.Slice` and add a buffer control

File: RabbitMQ.Stream.Client/Deliver.cs [hunk 2]
```diff
@@ -148,15 +154,27 @@ internal static int Read(ReadOnlySequence<byte> frame, out Chunk chunk)
...
             offset += WireFormatting.ReadUInt64(ref reader, out var chunkId);
-            offset += WireFormatting.ReadInt32(ref reader, out var crc);
+            offset += WireFormatting.ReadUInt32(ref reader, out var crc);
             offset += WireFormatting.ReadUInt32(ref reader, out var dataLen);
             offset += WireFormatting.ReadUInt32(ref reader, out _);
             // offset += 4; // reserved
             offset += WireFormatting.ReadUInt32(ref reader, out _); // reserved
             var memory =
                 ArrayPool<byte>.Shared.Rent((int)dataLen).AsMemory(0, (int)dataLen);
-            var data = reader.Sequence.Slice(reader.Consumed, dataLen);
```
Comment (Gsantomaggio) [inline RabbitMQ.Stream.Client/Deliver.cs:158]: read https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/285/files#r1317311527

**Review by lukebakken — COMMENTED** [review]
Rather than code and maintain our own Crc32 implementation, should we allow the user to provide one if they want this feature? i.e. if you want CRC32 checking, you must include one of these libraries with your application -
* https://github.com/force-net/Crc32.NET
* https://github.com/icsharpcode/SharpZipLib
* https://www.nuget.org/packages/System.IO.Hashing/6.0.0

**Comment by Gsantomaggio** [comment]
The idea was to reduce as much as possible the external dependencies.
If we want to add a library I would add [ttps://www.nuget.org/packages/System.IO.Hashing/6.0.0](https://www.nuget.org/packages/System.IO.Hashing/6.0.0)

File: RabbitMQ.Stream.Client/RabbitMQ.Stream.Client.csproj [hunk 3]
```diff
@@ -45,6 +45,7 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
       <PrivateAssets>all</PrivateAssets>
     </PackageReference>
+    <PackageReference Include="System.IO.Hashing"/>
```
Comment (Gsantomaggio) [inline RabbitMQ.Stream.Client/RabbitMQ.Stream.Client.csproj]: @lukebakken done!

File: RabbitMQ.Stream.Client/Deliver.cs (hunk 1 above)
Comment (lukebakken) [reply RabbitMQ.Stream.Client/Deliver.cs:98]: 👍

File: RabbitMQ.Stream.Client/PublicAPI.Unshipped.txt [hunk 4]
```diff
@@ -7,6 +7,7 @@ RabbitMQ.Stream.Client.AuthMechanism.External = 1 -> RabbitMQ.Stream.Client.Auth
 RabbitMQ.Stream.Client.AuthMechanism.Plain = 0 -> RabbitMQ.Stream.Client.AuthMechanism
 RabbitMQ.Stream.Client.AuthMechanismNotSupportedException
 RabbitMQ.Stream.Client.AuthMechanismNotSupportedException.AuthMechanismNotSupportedException(string s) -> void
+RabbitMQ.Stream.Client.Chunk.Crc.get -> uint
```
Comment (lukebakken) [inline RabbitMQ.Stream.Client/PublicAPI.Unshipped.txt:10]: Unfortunately this is a breaking change in the public API, because a library user who upgrades will get this when they compile:

```
error CS0266: Cannot implicitly convert type 'int' to 'uint'. An explicit conversion exists (are you missing a cast?)
```

...and I'm sure some sort of nasty exception at run time.

Is this change necessary?

**Review by lukebakken — CHANGES_REQUESTED** [review]

File: RabbitMQ.Stream.Client/PublicAPI.Unshipped.txt (hunk 4 above)
Comment (Gsantomaggio) [reply RabbitMQ.Stream.Client/PublicAPI.Unshipped.txt:10]: I had to change `  public readonly struct Deliver : ICommand` struct since the CRC coming from the server is a `uint` and not an `int`. 

Users don't use this class directly in fact I think the class `Deliver` as many others should per internal/private.

**Review by lukebakken — APPROVED** [review]
Thanks @Gsantomaggio !!!
